### PR TITLE
Passing pip install options

### PR DESCRIPTION
**Pull request by igor-davidyuk**
Workspace export routine is now aware of pip install options in requirements files.
Users may now pass options for remote pip install execution.

### CODE REVIEW

File: openfl/interface/workspace.py [hunk 1]
```diff
@@ -109,7 +110,12 @@ def create(prefix, template):
 
 
 @workspace.command(name='export')
-def export_():
+@option('-o', '--pip-install-options', required=False,
+        type=str, multiple=True, default=tuple,
+        help='Options for remote pip install. '
+        'You may pass several options in quotation marks alongside with arguments, '
```
Comment (aleksandr-mokrov) [inline openfl/interface/workspace.py]: Can you align text to text (Second and third row of help text under first row, not under argument `help`)?

File: openfl/interface/workspace.py [hunk 2]
```diff
@@ -124,19 +130,14 @@ def export_():
...
         echo(f'Plan file "{plan_file}" not found. No freeze performed.')
 
-    from pip._internal.operations import freeze
-    requirements_generator = freeze.freeze()
-    with open('./requirements.txt', 'w') as f:
-        for package in requirements_generator:
-            if is_package_versioned(package):
-                f.write(package + '\n')
+    # Dump requirements.txt
+    dump_requirements_file(prefixes=pip_install_options, keep_original_prefixes=True)
```
Comment (aleksandr-mokrov) [inline openfl/interface/workspace.py:140]: dump_requirements_file reads requirements.txt file to get prefixes form it and then rewrite that file by merged prefixes (pip_install_options with prefixes from requirements.txt) and installed packages (from `pip freeze`). Probably we should explicitly use two different file, one as a source of prefixes  and write to other?

File: openfl/interface/interactive_api/experiment.py [hunk 3]
```diff
@@ -149,15 +151,42 @@ def prepare_workspace_distribution(self, model_provider, task_keeper, data_loade
...
+        opt_treatment - Optimizer state treatment policy.
+            Valid options: 'RESET' - reinitialize optimizer for every round,
+            'CONTINUE_LOCAL' - keep local optimizer state,
+            'CONTINUE_GLOBAL' - aggregate optimizer state.
+        device_assignment_policy - device assignment policy.
+            Valid options: 'CPU_ONLY' - device parameter passed to tasks
+            will always be 'cpu',
+            'CUDA_PREFERRED' - enable passing CUDA device identifiers to tasks
+            by collaborators, works with cuda-device-monitor plugin equipped Envoys.
+        pip_install_options - list of options for the remote `pip install` calls,
```
Comment (aleksandr-mokrov) [inline openfl/interface/interactive_api/experiment.py]: It's a tuple

Comment (igor-davidyuk) [inline openfl/interface/interactive_api/experiment.py]:
```suggestion
        pip_install_options - tuple of options for the remote `pip install` calls,
```

File: openfl/interface/workspace.py [hunk 4]
```diff
@@ -109,7 +110,12 @@ def create(prefix, template):
 
 
 @workspace.command(name='export')
-def export_():
+@option('-o', '--pip-install-options', required=False,
+        type=str, multiple=True, default=tuple,
+        help='Options for remote pip install. '
+        'You may pass several options in quotation marks alongside with arguments, '
+        'e.g. -o "--find-links source.site"')
```
Comment (igor-davidyuk) [inline openfl/interface/workspace.py]:
```suggestion
        help='Options for remote pip install. '
          'You may pass several options in quotation marks alongside with arguments, '
          'e.g. -o "--find-links source.site"')
```

File: openfl/interface/workspace.py (hunk 1 above)
Comment (igor-davidyuk) [reply openfl/interface/workspace.py]: flake8:
./openfl/interface/workspace.py:116:11: E127 continuation line over-indented for visual indent

File: openfl/interface/interactive_api/experiment.py (hunk 3 above)
Comment (igor-davidyuk) [reply openfl/interface/interactive_api/experiment.py]: fixed

File: openfl/interface/workspace.py (hunk 2 above)
Comment (igor-davidyuk) [reply openfl/interface/workspace.py:140]: we do not manage different versions of requirements files

File: openfl/interface/workspace.py [hunk 5]
```diff
@@ -109,7 +110,12 @@ def create(prefix, template):
 
 
 @workspace.command(name='export')
-def export_():
+@option('-o', '--pip-install-options', required=False,
+        type=str, multiple=True, default=tuple,
+        help='Options for remote pip install. '
+          'You may pass several options in quotation marks alongside with arguments, '
+          'e.g. -o "--find-links source.site"')
```
Comment (igor-davidyuk) [inline openfl/interface/workspace.py]:
```suggestion
        'You may pass several options in quotation marks alongside with arguments, '
        'e.g. -o "--find-links source.site"')
```

File: openfl/utilities/workspace.py [hunk 6]
```diff
@@ -75,3 +78,52 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+    if keep_original_prefixes and path.is_file():
+        with open(path) as f:
+            while True:
+                line = f.readline()
+                if line == '':
+                    continue
+                if line[0] == '-':
+                    prefixes |= {line.replace('\n', '')}
+                else:
+                    break
```
Comment (dmitryagapov) [inline openfl/utilities/workspace.py]:
```suggestion
            for line in f:
                if line[0] == '-':
                    prefixes |= {line.replace('\n', '')}
                else:
                    break
```

File: tests/openfl/utilities/test_dump_requirements.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    expected_options = []
+    for li in expected_lines:
+        if li[0] == '-':
+            expected_options.append(li)
+        else:
+            break
+    expected_packages = expected_lines[len(expected_options):]
+
+    assert len(read_options) == len(expected_options)
+    assert len(read_packages) == len(expected_packages)
```
Comment (dmitryagapov) [inline tests/openfl/utilities/test_dump_requirements.py:70]: This is not required.

File: openfl/interface/workspace.py (hunk 1 above)
Comment (aleksandr-mokrov) [reply openfl/interface/workspace.py]: If you add only tab, then it will be that error, but if you align text to text ('You ...' strictly under 'Options...' ). Like this https://github.com/intel/openfl/blob/2d66cd816a78670d6a99ef04669f5fa47bcf45f7/openfl/interface/collaborator.py#L255

File: tests/openfl/utilities/test_dump_requirements.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    monkeypatch.setattr(freeze, 'freeze', mock_pip_freeze)
+
+    dump_requirements_file(path=requirements_file,
+                           keep_original_prefixes=keep_original_prefixes,
+                           prefixes=prefixes)
+
+    with open(requirements_file) as f:
+        read_lines = f.readlines()
+
+    read_options = []
```
Comment (dmitryagapov) [inline tests/openfl/utilities/test_dump_requirements.py:53]: I don't think that `read_options` and `expected_options` are the good naming

File: tests/openfl/utilities/test_dump_requirements.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+@pytest.mark.parametrize('keep_original_prefixes,prefixes,expected_lines', [
+    (False, None, LINES[1:]),
+    (True, None, LINES),
+    (False, PREFIXES, [p + '\n' for p in PREFIXES] + [li for li in LINES if li[0] != '-']),
+    (True, PREFIXES, [li for li in LINES if li[0] == '-'
+                      ] + [p + '\n' for p in PREFIXES] + [li for li in LINES if li[0] != '-']),
+    (True, PREFIXES_OVERLAP, [p + '\n' for p in PREFIXES_OVERLAP
+                              ] + [li for li in LINES if li[0] != '-']),
+])
```
Comment (dmitryagapov) [inline tests/openfl/utilities/test_dump_requirements.py]: To simplify it, maybe make sense:
@pytest.mark.parametrize('keep_original_prefixes,prefixes,expected_lines', [
    (False, None, REQ_WITHOUT_PREFIXES),
    (True, None, REQ_WITH_PREFIXES),
    (False, PREFIXES, PREFIXES + REQ_WITHOUT_PREFIXES,
    (True, PREFIXES, PRESIXES + REQ_WITH_PREFIXES,
    (True, PREFIXES_OVERLAP, PREFIXES_OVERLAP + REQ_WITH_PREFIXES),
])
Because multiple list comprehension hard to read.
Also maybe this test should be separated to increase readability.

**Review by dmitryagapov — CHANGES_REQUESTED** [review]

**Comment by dmitryagapov** [comment]
I think that we need to test such cases as requirements.txt doesn't have prefixes and empty requirements.txt

File: tests/openfl/utilities/test_dump_requirements.py (hunk 7 above)
Comment (igor-davidyuk) [reply tests/openfl/utilities/test_dump_requirements.py:70]: why?

File: tests/openfl/utilities/test_dump_requirements.py (hunk 8 above)
Comment (igor-davidyuk) [reply tests/openfl/utilities/test_dump_requirements.py:53]: Do you have any suggestions?

File: tests/openfl/utilities/test_dump_requirements.py (hunk 9 above)
Comment (igor-davidyuk) [reply tests/openfl/utilities/test_dump_requirements.py]: These list comprehension will be hard to read but in another place

File: openfl/utilities/workspace.py (hunk 6 above)
Comment (igor-davidyuk) [reply openfl/utilities/workspace.py]: replaced while with for

File: tests/openfl/utilities/test_dump_requirements.py (hunk 7 above)
Comment (dmitryagapov) [reply tests/openfl/utilities/test_dump_requirements.py:70]: Because `set(read_options) == set(expected_options)` can't be equal if its have different length

Comment (igor-davidyuk) [reply tests/openfl/utilities/test_dump_requirements.py:70]: x = ['a', 'a', 'b']
y = ['a', 'b']
set(x) == set(y)  # True

File: tests/openfl/utilities/test_dump_requirements.py (hunk 9 above)
Comment (igor-davidyuk) [reply tests/openfl/utilities/test_dump_requirements.py]: Replaced list comprehension with variables

**Comment by igor-davidyuk** [comment]
> I think that we need to test such cases as requirements.txt doesn't have prefixes and empty requirements.txt

Added additional tests

File: tests/openfl/utilities/test_dump_requirements.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+def test_dump_empty_original_list(
+        monkeypatch, touch_file,
+        keep_original_prefixes, prefixes, expected_lines):
+    """Test dump_requirements_file function with no file to start."""
+    def mock_pip_freeze():
+        return [line.replace('\n', '') for line in ORIGINAL_PACKAGE_LINES]
+    monkeypatch.setattr(freeze, 'freeze', mock_pip_freeze)
+
+    requirements_file = Path('./test_requirements_2.txt').absolute()
```
Comment (aleksandr-mokrov) [inline tests/openfl/utilities/test_dump_requirements.py:95]: You can use fixture like` requirements_file` but with args to pass `touch_file`, or tmp_path to be sure that file will be unlinked if test fails with exception before `requirements_file.unlink()`.

**Review by aleksandr-mokrov — APPROVED** [review]

File: openfl/utilities/workspace.py [hunk 11]
```diff
@@ -75,3 +78,51 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+        prefixes = set(prefixes)
+
+    # Merge prefixes:
+    # We expect that all the prefixes in a requirement file
+    # are placed at the top
+    if keep_original_prefixes and path.is_file():
+        with open(path) as f:
+            for line in f:
+                if line == '':
+                    continue
```
Comment (dmitryagapov) [inline openfl/utilities/workspace.py]: Looks like it is not needed. Or maybe you found corner case for it?

**Review by dmitryagapov — APPROVED** [review]

File: tests/openfl/utilities/test_dump_requirements.py (hunk 10 above)
Comment (igor-davidyuk) [reply tests/openfl/utilities/test_dump_requirements.py:95]: I believe parametrizing the fixture will lead to overengineering this test module. The problem is we can not broadcast expected results to different values of the `touch_file` parameter.
Instead, I introduced `try - finally` construction to make sure the `test_requirements` file is always unlinked.

File: openfl/utilities/workspace.py (hunk 11 above)
Comment (igor-davidyuk) [reply openfl/utilities/workspace.py]: This is a mistake. I fixed it to state
```
if line == '\n':
    continue
```
The idea is to skip empty lines between options in the original requirements.txt, for instance:
```
-o option1

-s option2
requirement1==0.0.1
```

File: tests/openfl/utilities/test_dump_requirements.py (hunk 8 above)
Comment (dmitryagapov) [reply tests/openfl/utilities/test_dump_requirements.py:53]: I thought that `read` is verb)

**Review by alexey-gruzdev — APPROVED** [review]
